### PR TITLE
fix: show actionable dialog when Steam session is taken by another device

### PR DESCRIPTION
**Pull request by ben-pearson**
## Problem

When a PC starts a Steam game while a game is already running in Wine on the device, Steam sends `EResult.LoggedInElsewhere` inside `LoggedOffCallback`. The previous code treated this the same as any other logoff — it emitted `ForceCloseApp` and silently killed the game with no feedback to the user.

https://github.com/user-attachments/assets/0dbe8ceb-64e6-45e8-ba8c-316e5e055e96

**Root cause confirmed via logcat:** `PlayingSessionStateCallback` was assumed to fire in this scenario, but it never does. `LoggedInElsewhere` inside `LoggedOffCallback` is the only signal Steam sends.

## Changes

https://github.com/user-attachments/assets/a31124f0-59d1-44f6-b4f0-4ff3198286df

**`SteamEvent.kt`** — adds `PlayingBlocked` event to signal the conflict to the UI layer.

**`SteamService.kt`**
- `onLoggedOff`: when `LoggedInElsewhere` arrives and a game is running (`xEnvironment != null`), show the conflict dialog instead of force-closing. When no game is running, the existing `ForceCloseApp` behaviour is preserved.
- `_isHandlingConflict` (`AtomicBoolean`): prevents the dialog being shown multiple times if `LoggedInElsewhere` fires repeatedly during reconnect attempts.
- `clearPlayingConflict()`: resets conflict state on game exit.
- `reconnect()` / `onDisconnected()`: suppress `Disconnected`/`RemotelyDisconnected` events while a conflict is being handled, so the reconnect cycle doesn't flash a "Disconnected from Steam" banner behind the dialog.
- `onPlayingSessionState`: safety net — also emits `PlayingBlocked` if `PlayingSessionStateCallback` fires with `isPlayingBlocked=true` while in-game.

**`PluviaApp.kt`** — calls `SteamService.clearPlayingConflict()` in `shutdownEnvironment()` to reset state when the game exits.

**`XServerScreen.kt`** — registers a listener for `SteamEvent.PlayingBlocked` and shows an `AlertDialog` with:
- **Play Anyway** — calls `kickPlayingSession(onlyGame=true)` to stop the PC's game session
- **Cancel** — exits the local game cleanly

`onDismissRequest` is intentionally empty — the user must make an explicit choice.

**`strings.xml`** — adds `main_app_running_unknown_game` ("another device"), used as the fallback game name in the dialog message when the conflicting game title cannot be determined.

## Note

Behaviour varies slightly depending on the conflicting device (PC, Steam Deck, Mac). This has been tested across all three and is improved in each case, but given the variance in how Steam signals session conflicts across platforms, further edge cases may exist.

## Test plan
- [x] Launch a game on PC, then launch a game on device — pre-launch "App Running" dialog appears (existing behaviour, unaffected)
- [x] Launch a game on device, then start a game on PC — in-game "App Running" dialog appears without a "Disconnected from Steam" banner
- [x] Press **Cancel** — game exits cleanly
- [x] Press **Play Anyway** — PC game session is kicked, device continues playing
- [x] Normal game exit with no conflict — no regression to shutdown flow

---
## Summary by cubic
Show an actionable “App Running” dialog when Steam logs the user off because another device started a game, instead of silently closing the local game. The dialog persists across Activity recreation and hides the disconnect banner; users can kick the remote session or exit.

- **Bug Fixes**
  - Handle `EResult.LoggedInElsewhere` in `SteamService.onLoggedOff`: if a game is running, emit `PlayingBlocked` (not `ForceCloseApp`); otherwise keep existing behavior.
  - Show an `AlertDialog` in `XServerScreen` on `PlayingBlocked` with “Play anyway” (call `SteamService.clearPlayingConflict()` then `SteamService.kickPlayingSession(onlyGame=true)`) and “Cancel” (clean exit); persist with `rememberSaveable` and safe-call `xServerView`.
  - Add `_isHandlingConflict` to prevent duplicate dialogs; suppress `Disconnected`/`RemotelyDisconnected` while handling; reset via `SteamService.clearPlayingConflict()` on shutdown and before kicking.
  - Also emit `PlayingBlocked` when `PlayingSessionStateCallback.isPlayingBlocked` is true; add `main_app_running_unknown_game` fallback and translations across locales.

<sup>Written for commit b9083cf285fadcb4836f9fa8f1c17df6ea7fe47f. Summary will update on new commits. <a href="https://cubic.dev/pr/utkarshdalal/GameNative/pull/1306?utm_source=github">Review in cubic</a></sup>

## Summary by CodeRabbit

* **New Features**
  * Detects when a game session is blocked by another device and notifies the app.
  * Shows a conflict-resolution dialog to stop the remote session or exit the current session.
  * Adds a fallback label when the remote game name is unknown.

* **Bug Fixes**
  * Improved cleanup/reset of Steam-related session/conflict state during app shutdown.

### CODE REVIEW

File: app/src/main/java/app/gamenative/service/SteamService.kt [hunk 1]
```diff
@@ -417,6 +418,11 @@ class SteamService : Service(), IChallengeUrlChanged {
         var isWaitingForQRAuth: Boolean = false
             private set
 
+        fun clearPlayingConflict() {
+            instance?._isPlayingBlocked?.value = false
+            instance?._isHandlingConflict?.set(false)
+        }
```
[the remark this answers is not in the document]
Comment (ben-pearson) [reply app/src/main/java/app/gamenative/service/SteamService.kt:424]: Fixed in ed16fa5 : `clearPlayingConflict()` is now called synchronously in the Play Anyway onClick handler before the kick coroutine is launched, so `_isHandlingConflict` is always reset regardless of kick outcome.

File: app/src/main/java/app/gamenative/service/SteamService.kt [hunk 2]
```diff
@@ -3538,6 +3556,10 @@ class SteamService : Service(), IChallengeUrlChanged {
     private fun onPlayingSessionState(callback: PlayingSessionStateCallback) {
         Timber.d("onPlayingSessionState called with isPlayingBlocked = " + callback.isPlayingBlocked)
         _isPlayingBlocked.value = callback.isPlayingBlocked
+        if (callback.isPlayingBlocked) {
+            val event = SteamEvent.PlayingBlocked
+            PluviaApp.events.emit(event)
+        }
```
[the remark this answers is not in the document]
Comment (ben-pearson) [reply app/src/main/java/app/gamenative/service/SteamService.kt]: Fixed : `onPlayingSessionState` now uses `compareAndSet(false, true)` consistent with the `onLoggedOff` path, so the safety-net only fires when no conflict is already being handled.

File: app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt [hunk 3]
```diff
@@ -2245,6 +2254,32 @@ fun XServerScreen(
...
+            confirmButton = {
+                TextButton(onClick = {
+                    showPlayingBlockedDialog = false
+                    scope.launch {
+                        SteamService.kickPlayingSession(onlyGame = true)
+                    }
+                }) {
+                    Text(text = stringResource(R.string.main_play_anyway))
+                }
+            },
```
[the remark this answers is not in the document]
Comment (ben-pearson) [reply app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt:2272]: Fixed in ed16fa5 : `clearPlayingConflict()` is now called synchronously before `scope.launch`, so conflict state is always restored immediately regardless of kick outcome. Note: `kickPlayingSession` always times out in the `LoggedInElsewhere` scenario (the `PlayingSessionStateCallback` never fires), but the kick itself does work - awaiting the result would block the UI for 5 seconds with no benefit.

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -1065,6 +1065,7 @@
     <string name="main_app_running_message">You are logged in on another device already playing %s. \nYou can still play this game, but that will disconnect the other session from Steam.</string>
     <string name="main_app_running_other_device">You are logged in on another device (%1$s) already playing %2$s (%3$s), and that save is not yet in the cloud. \nYou can still play this game, but that will disconnect the other session from Steam and may create a save conflict when that session progress is synced</string>
     <string name="main_play_anyway">Play anyway</string>
+    <string name="main_app_running_unknown_game">another device</string>
```
[the remark this answers is not in the document]
Comment (ben-pearson) [reply app/src/main/res/values/strings.xml]: Fixed in ed16fa5 — `clearPlayingConflict()` is now called synchronously before the kick coroutine in the Play Anyway handler.

File: app/src/main/java/app/gamenative/service/SteamService.kt [hunk 5]
```diff
@@ -3526,10 +3536,18 @@ class SteamService : Service(), IChallengeUrlChanged {
         } else if (callback.result == EResult.LoggedInElsewhere) {
             // received when a client runs an app and wants to forcibly close another
             // client running an app
-            val event = SteamEvent.ForceCloseApp
-            PluviaApp.events.emit(event)
-
-            reconnect()
+            if (PluviaApp.xEnvironment != null) {
+                if (!_isHandlingConflict.getAndSet(true)) {
```
[the remark this answers is not in the document]
Comment (ben-pearson) [reply app/src/main/java/app/gamenative/service/SteamService.kt:3540]: Fixed in ed16fa5 : `clearPlayingConflict()` is called synchronously in the Play Anyway onClick before the kick coroutine, resetting both `_isPlayingBlocked` and `_isHandlingConflict` immediately.

File: app/src/main/java/app/gamenative/service/SteamService.kt [hunk 6]
```diff
@@ -417,6 +418,11 @@ class SteamService : Service(), IChallengeUrlChanged {
         var isWaitingForQRAuth: Boolean = false
             private set
 
+        fun clearPlayingConflict() {
```
[the remark this answers is not in the document]
Comment (ben-pearson) [reply app/src/main/java/app/gamenative/service/SteamService.kt:421]: Fixed in ed16fa5 : `clearPlayingConflict()` is now called synchronously before `scope.launch` in the Play Anyway handler, so `_isHandlingConflict` is always reset regardless of kick result.

File: app/src/main/res/values/strings.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (ben-pearson) [reply app/src/main/res/values/strings.xml]: Fixed : the string resource value was changed from "another device" to "a game", so the dialog now reads naturally (e.g. "already playing a game").

File: app/src/main/java/app/gamenative/service/SteamService.kt [hunk 7]
```diff
@@ -3538,6 +3556,10 @@ class SteamService : Service(), IChallengeUrlChanged {
     private fun onPlayingSessionState(callback: PlayingSessionStateCallback) {
         Timber.d("onPlayingSessionState called with isPlayingBlocked = " + callback.isPlayingBlocked)
         _isPlayingBlocked.value = callback.isPlayingBlocked
+        if (callback.isPlayingBlocked) {
```
[the remark this answers is not in the document]
Comment (ben-pearson) [reply app/src/main/java/app/gamenative/service/SteamService.kt]: Fixed : `onPlayingSessionState` now uses `compareAndSet(false, true)` to gate the emit, consistent with the `onLoggedOff` path.

File: app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt [hunk 8]
```diff
@@ -2245,6 +2254,32 @@ fun XServerScreen(
...
     }
 
+    if (showPlayingBlockedDialog) {
+        androidx.compose.material3.AlertDialog(
+            onDismissRequest = {},
+            title = { Text(text = stringResource(R.string.main_app_running_title)) },
+            text = { Text(text = stringResource(R.string.main_app_running_message, context.getString(R.string.main_app_running_unknown_game))) },
+            confirmButton = {
+                TextButton(onClick = {
+                    showPlayingBlockedDialog = false
```
[the remark this answers is not in the document]
Comment (ben-pearson) [reply app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt:2264]: Fixed in ed16fa5 : `clearPlayingConflict()` is now called synchronously before `scope.launch`. Note: `kickPlayingSession` always times out (5 s) in this scenario because `PlayingSessionStateCallback` never fires when the trigger is `LoggedInElsewhere`; the kick itself does succeed. Awaiting the result would block the UI for 5 seconds with no actionable outcome, so we restore clean state immediately instead.

File: app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt [hunk 9]
```diff
@@ -411,6 +413,7 @@ fun XServerScreen(
     var showElementEditor by remember { mutableStateOf(false) }
     var elementToEdit by remember { mutableStateOf<com.winlator.inputcontrols.ControlElement?>(null) }
     var showPhysicalControllerDialog by remember { mutableStateOf(false) }
+    var showPlayingBlockedDialog by remember { mutableStateOf(false) }
```
[the remark this answers is not in the document]
Comment (ben-pearson) [reply app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt]: Fixed in 1d15369 — changed to `rememberSaveable` so the dialog survives Activity recreation while a conflict is active.

File: app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt [hunk 10]
```diff
@@ -2245,6 +2254,33 @@ fun XServerScreen(
...
+                    }
+                }) {
+                    Text(text = stringResource(R.string.main_play_anyway))
+                }
+            },
+            dismissButton = {
+                TextButton(onClick = {
+                    showPlayingBlockedDialog = false
+                    exit(xServerView!!.getxServer().winHandler, frameRating, currentAppInfo, container, appId, onExit, navigateBack)
+                }) {
```
[the remark this answers is not in the document]
Comment (ben-pearson) [reply app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt:2277]: Fixed in 874a0d6 - changed to `xServerView?.getxServer()?.winHandler` to avoid an NPE if Cancel is tapped before xServerView is rebound after Activity recreation. The `exit()` function already accepts a nullable WinHandler so no further changes were needed.

File: app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt [hunk 11]
```diff
@@ -411,6 +413,7 @@ fun XServerScreen(
     var showElementEditor by remember { mutableStateOf(false) }
     var elementToEdit by remember { mutableStateOf<com.winlator.inputcontrols.ControlElement?>(null) }
     var showPhysicalControllerDialog by remember { mutableStateOf(false) }
+    var showPlayingBlockedDialog by rememberSaveable { mutableStateOf(false) }
```
[the remark this answers is not in the document]
Comment (ben-pearson) [reply app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt:416]: Fixed in 874a0d6 - same as above, changed to `xServerView?.getxServer()?.winHandler`.

File: app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt [hunk 12]
```diff
@@ -2245,6 +2254,33 @@ fun XServerScreen(
...
+            onDismissRequest = {},
+            title = { Text(text = stringResource(R.string.main_app_running_title)) },
+            text = { Text(text = stringResource(R.string.main_app_running_message, context.getString(R.string.main_app_running_unknown_game))) },
+            confirmButton = {
+                TextButton(onClick = {
+                    showPlayingBlockedDialog = false
+                    SteamService.clearPlayingConflict()
+                    scope.launch {
+                        SteamService.kickPlayingSession(onlyGame = true)
+                    }
```
[the remark this answers is not in the document]
Comment (ben-pearson) [reply app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt:2268]: Not changing this. `kickPlayingSession` always times out after 5s in the `LoggedInElsewhere` scenario because `PlayingSessionStateCallback` never fires - it is confirmed via logcat that Steam does not emit this callback when a PC game takes over the session. The kick itself does work despite the timeout. Clearing conflict state before the coroutine means disconnect events are unsuppressed immediately on dismiss, rather than leaving the user with a 5s window where a genuine disconnect would be silently swallowed.

File: app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt [hunk 13]
```diff
@@ -2245,6 +2254,33 @@ fun XServerScreen(
...
+            dismissButton = {
+                TextButton(onClick = {
+                    showPlayingBlockedDialog = false
+                    exit(xServerView?.getxServer()?.winHandler, frameRating, currentAppInfo, container, appId, onExit, navigateBack)
+                }) {
+                    Text(text = stringResource(R.string.cancel))
+                }
+            },
+        )
+    }
```
Comment (utkarshdalal) [inline app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt:2282]: Curious why you decided to put the dialog here instead of PluviaMain with the others? Is it required to be here?

File: app/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -1065,6 +1065,7 @@
     <string name="main_app_running_message">You are logged in on another device already playing %s. \nYou can still play this game, but that will disconnect the other session from Steam.</string>
     <string name="main_app_running_other_device">You are logged in on another device (%1$s) already playing %2$s (%3$s), and that save is not yet in the cloud. \nYou can still play this game, but that will disconnect the other session from Steam and may create a save conflict when that session progress is synced</string>
     <string name="main_play_anyway">Play anyway</string>
+    <string name="main_app_running_unknown_game">a game</string>
```
Comment (utkarshdalal) [inline app/src/main/res/values/strings.xml:1068]: translations pls

**Comment by utkarshdalal** [comment]
couple of comments but otherwise looks good

File: app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt (hunk 13 above)
Comment (ben-pearson) [reply app/src/main/java/app/gamenative/ui/screen/xserver/XServerScreen.kt:2282]: the Cancel handler needs to call exit() which handles Wine process teardown via winHandler - that context only exists in XServerScreen, is there a better way to do this? I was trying to change as little as possible (reducing risk)

File: app/src/main/res/values/strings.xml (hunk 14 above)
Comment (ben-pearson) [reply app/src/main/res/values/strings.xml:1068]: added (AI generated translations) hope they are correct :-D 🏴󠁧󠁢󠁥󠁮󠁧󠁿
